libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts

```diff
@@ -0,0 +1,167 @@
+ * For a distributed bot in production, this requires a distributed storage
+ * ensuring only one token exchange is processed. This middleware supports
+ * CosmosDb storage found in Microsoft.Bot.Builder.Azure, or MemoryStorage for
+ * local development. IStorage's ETag implementation for token exchange activity
+ * deduplication.
+ */
+export class TeamsSSOTokenExchangeMiddleware implements Middleware {
+    /**
+     * Initializes a new instance of the TeamsSSOTokenExchangeMiddleware class.
+     *
+     * @param storage The [Storage](xref:botbuilder-core.Storage) to use for deduplication
+     * @param oAuthConnectionName The connection name to use for the single sign on token exchange
+     */
+    constructor(private readonly storage: Storage, private readonly oAuthConnectionName: string) {}
+
+    /**
+     * Called each time the bot receives a new request.
+     *
+     * @param context Context for current turn of conversation with the user.
+     * @param next Function to call to continue execution to the next step in the middleware chain.
+     */
+    async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
+        if (context.activity.name === signInConstants.tokenExchangeOperationName) {
+            // If the TokenExchange is NOT successful, the response will have already been sent by ExchangedTokenAsync
+            if (!(await this.exchangedToken(context))) {
+                return;
+            }
+
+            // Only one token exchange should proceed from here. Deduplication is performed second because in the case
+            // of failure due to consent required, every caller needs to receive the
+            if (!(await this.deduplicatedTokenExchangeId(context))) {
+                // If the token is not exchangeable, do not process this activity further.
+                return;
+            }
+        }
+
+        await next();
+    }
+
+    private async deduplicatedTokenExchangeId(context: TurnContext): Promise<boolean> {
+        // Create a StoreItem with Etag of the unique 'signin/tokenExchange' request
+        const storeItem: StoreItem = {
+            eTag: context.activity.value?.id,
+        };
+
+        try {
+            // Writing the IStoreItem with ETag of unique id will succeed only once
+            await this.storage.write({
+                [getStorageKey(context)]: storeItem,
+            });
+        } catch (err) {
+            const message = err.message?.toLowerCase();
+
+            // Do NOT proceed processing this message, some other thread or machine already has processed it.
+            // Send 200 invoke response.
+            if (message.includes('etag conflict') || message.includes('precondition is not met')) {
+                await this.sendInvokeResponse(context);
+                return false;
+            }
+
+            throw err;
+        }
+
+        return true;
+    }
+
+    private async sendInvokeResponse(
```

Comment:
Could be static or internal (not PR blocking)

Reply:
Good suggestion!